examples/sui-move/escrow/sources/shared_example.move

```diff
@@ -0,0 +1,290 @@
+        sender: address,
+        /// intended recipient of the escrowed object
+        recipient: address,
+        /// ID of the object `sender` wants in exchange
+        // TODO: this is probably a bad idea if the object is mutable.
+        // that can be fixed by asking for an additional approval
+        // from `sender`, but let's keep it simple for now.
+        exchange_for: ID,
+        /// the escrowed object that we store into an option because it could already be taken
+        escrowed: Option<T>,
+    }
+
+    // Error codes
+    /// The `sender` and `recipient` of the two escrowed objects do not match
+    const EMismatchedSenderRecipient: u64 = 0;
+    /// The `exchange_for` fields of the two escrowed objects do not match
+    const EMismatchedExchangeObject: u64 = 1;
+    /// The escrow has already been exchanged or cancelled
+    const EAlreadyExchangedOrCancelled: u64 = 3;
+
+    public fun create<T: key + store>(
+        recipient: address,
+        exchange_for: ID,
+        escrowed_item: T,
+        ctx: &mut TxContext
+    ) {
+        let sender = tx_context::sender(ctx);
+        let id = object::new(ctx);
+        let escrowed = option::some(escrowed_item);
+
+        transfer::public_share_object(
+            Exchange<T> {
+                id, sender, recipient, exchange_for, escrowed
+            }
+        );
+    }
+
+    /// The `recipient` of the escrow can exchange `obj` with the escrowed item
+    public fun exchange<T: key + store, ExchangeForT: key + store>(
+        obj: ExchangeForT,
+        escrow: &mut Exchange<T>,
+        ctx: &TxContext
+    ) {
+        assert!(option::is_some(&escrow.escrowed), EAlreadyExchangedOrCancelled);
+        let escrowed_item = option::extract<T>(&mut escrow.escrowed);
+        assert!(&tx_context::sender(ctx) == &escrow.recipient, EMismatchedSenderRecipient);
+        assert!(object::borrow_id(&obj) == &escrow.exchange_for, EMismatchedExchangeObject);
+        // everything matches. do the swap!
+        transfer::public_transfer(escrowed_item, tx_context::sender(ctx));
+        transfer::public_transfer(obj, escrow.sender);
+    }
+
+    /// The `creator` can cancel the escrow and get back the escrowed item
+    public fun cancel<T: key + store>(
+        escrow: &mut Exchange<T>,
+        ctx: &TxContext
+    ): T {
+        assert!(&tx_context::sender(ctx) == &escrow.sender, EMismatchedSenderRecipient);
+        assert!(option::is_some(&escrow.escrowed), EAlreadyExchangedOrCancelled);
+        option::extract<T>(&mut escrow.escrowed)
+    }
+
+    // === Tests ===
+    use sui::coin::{Self, Coin};
+    use sui::sui::SUI;
+    #[test_only] use sui::test_scenario::{Self, Scenario};
+
+    #[test_only]
+    fun test_coin(ts: &mut Scenario): Coin<SUI> {
+        coin::mint_for_testing<SUI>(42, test_scenario::ctx(ts))
+    }
+
+    #[test]
+    fun test_successful_swap() {
+        let ts = test_scenario::begin(@0x0);
+        let alice = @0xA;
+        let bob = @0xB;
```

Comment:
Paying forward a good review comment from @manolisliolios, to factor these out into shared constants, so that you can use them in all your tests.